File: packages/utils/src/logger.zig
const std = @import("std");
const string = @import("./main.zig").string;

const Level = enum { info, warn, err };

const Color = struct {
    pub const Red = "\x1b[31m";
    pub const Green = "\x1b[32m";
    pub const Yellow = "\x1b[33m";
    pub const Reset = "\x1b[0m";
};

pub fn info(comptime message: string, args: anytype) void {
    log(.info, message, args);
}

pub fn infoNoBreak(comptime message: string, args: anytype) void {
    logNoBreak(.info, message, args);
}

pub fn warn(comptime message: string, args: anytype) void {
    log(.warn, message, args);
}

pub fn err(comptime message: string, args: anytype) void {
    log(.err, message, args);
}

fn log(level: Level, comptime message: string, args: anytype) void {
    const writer = std.io.getStdOut().writer();
    switch (level) {
        .info => {
            _ = writer.print(Color.Green, .{}) catch unreachable;
            _ = writer.print("info", .{}) catch unreachable;
        },
        .warn => {
            _ = writer.print(Color.Yellow, .{}) catch unreachable;
            _ = writer.print("warn", .{}) catch unreachable;
        },
        .err => {
            _ = writer.print(Color.Red, .{}) catch unreachable;
            _ = writer.print("error", .{}) catch unreachable;
        },
    }

    _ = writer.print(Color.Reset, .{}) catch unreachable;
    _ = writer.print(": ", .{}) catch unreachable;
    _ = writer.print(message ++ "\n", args) catch unreachable;
}

pub fn logNoBreak(level: Level, comptime message: string, args: anytype) void {
    const writer = std.io.getStdOut().writer();
    switch (level) {
        .info => {
            _ = writer.print(Color.Green, .{}) catch unreachable;
            _ = writer.print("info", .{}) catch unreachable;
        },
        .warn => {
            _ = writer.print(Color.Yellow, .{}) catch unreachable;
            _ = writer.print("warn", .{}) catch unreachable;
        },
        .err => {
            _ = writer.print(Color.Red, .{}) catch unreachable;
            _ = writer.print("error", .{}) catch unreachable;
        },
    }

    _ = writer.print(Color.Reset, .{}) catch unreachable;
    _ = writer.print(": ", .{}) catch unreachable;
    _ = writer.print(message, args) catch unreachable;
}

File: src/common/Logger.ts
enum Level {
  info = 'info',
  warn = 'warn',
  err = 'error',
  debug = 'debug',
}

const Color = {
  Red: '#E47B83',
  Green: '#98C379',
  Yellow: '#E5C07B',
  Debug: '#707782',
  Orange: '#f5c542',
  Reset: 'inherit',
};

export class Logger {
  public name: string;
  constructor(name: string) {
    this.name = name;
  }

  public info(...message: string[]) {
    this.log(Level.info, ...message);
  }

  public warn(...message: string[]) {
    this.log(Level.warn, ...message);
  }

  public err(...message: string[]) {
    this.log(Level.err, ...message);
  }

  public debug(...message: string[]) {
    this.log(Level.debug, ...message);
  }

  private log(level: Level, ...message: string[]) {
    let color;
    switch (level) {
      case Level.info:
        color = Color.Green;
        break;
      case Level.warn:
        color = Color.Yellow;
        break;
      case Level.debug:
        color = Color.Debug;
        break;
      default:
        color = Color.Red;
        break;
    }

    console.info(
      `%c Hykord %c %c ${this.name} %c`,
      `background: ${color}; color: black; font-weight: bold; border-radius: 5px;`,
      `color: ${Color.Reset}`,
      `background: ${Color.Orange}; color: black; font-weight: bold; border-radius: 5px;`,
      `color: ${Color.Reset}`,
      ...message,
    );
  }
}
